src/Network/DNS/FFI.hs

```diff
@@ -52,6 +54,20 @@ withCResState act
                          act ptr
   | otherwise = withMVar resolvLock $ \() -> act nullPtr
 
+withCResInit :: Ptr CResState -> IO a -> IO a
+withCResInit stptr act = bracket
+     (do
+         rc1 <- c_res_opt_set_use_dnssec stptr
+         unless (rc1 == 0) $
+             fail "res_init(3) failed"
+         return ()
+     )
+     (const $ c_res_close stptr)
+     (const $ do
+         resetErrno
```

Comment:
Could `resetErrno` be part of the initializer?  Seems more canonical...

Reply:
I think yes

Reply:
Then `resetErrno` could replace `return ()`.

```haskell
withCResInit :: Ptr CResState -> IO a -> IO a
withCResInit stptr act = bracket
     (do
         rc1 <- c_res_opt_set_use_dnssec stptr
         unless (rc1 == 0) $
             fail "res_init(3) failed"
         resetErrno
     )
     (const $ c_res_close stptr)
     (const act)
```

Reply:
Ok, did that, along with introducing local definitions to prevent long-ranging parentheses.

Reply:
Your rewrite can be further simplified (removing `act` binfdng):

```haskell
withCResInit :: Ptr CResState -> IO a -> IO a
withCResInit stptr = bracket initialize finalize . const
  where
    initialize = do
        rc1 <- c_res_opt_set_use_dnssec stptr
        unless (rc1 == 0) $
            fail "res_init(3) failed"
        resetErrno
    finalize _ = c_res_close stptr
```